src/engine/execute.js

```diff
@@ -21,9 +30,28 @@ var execute = function (sequencer, thread) {
     }
 
     var blockFunction = runtime.getOpcodeFunction(opcode);
-    if (!blockFunction) {
-        console.warn('Could not get implementation for opcode: ' + opcode);
-        return;
+    var isHat = runtime.getIsHat(opcode);
+
+    // Hats are implemented slightly differently from regular blocks.
+    // If they have an associated block function, it's treated as a predicate;
+    // if not, execution will proceed right through it (as a no-op).
+    if (isHat) {
+        var nextBlock = target.blocks.getNextBlock(currentBlockId);
+        if (!nextBlock) {
+            // Hat with no next block - don't try to evaluate it.
+            sequencer.retireThread(thread);
+            return;
```

Comment:
I definitely think @liam4's example should query the `pop` reporter, but I don't have a problem with it doing so ~30 times per second (or more) if that's the convenient implementation.

Hardware extension reporters frequently have side effects, at least at a small scale. For example, at the hardware level the PicoBoard only reports its sensor data (slider position, etc.) in response to a "get sensor data" command. One could implement a PicoBoard extension where each reporter is actually a requester which sends a "get sensor data" command, waits for the reply, processes the reply, and calls Scratch back with the result. If this were called frequently on a reasonably fast computer, the serial port's receive and maybe even transmit buffer could overflow, disrupting communication with the PicoBoard.

I think it's up to the extension author to avoid problems with calling their reporters frequently. In the PicoBoard example, we use `setInterval` to send the "get sensor data" command at a regular interval and cache the response. That way we can implement plain reporters (instead of requesters) in terms of that cache and we don't have to worry about a traffic jam at the serial port regardless of how frequently Scratch is querying the PicoBoard extension's reporters.